File: plex-now-playing.py
import os
import json
import requests
import re
from datetime import datetime
from xml.etree import ElementTree as ET
import keyring

WORKING_DIR = '~/scripts/'  # Working directory (ending with /) where script saves tokens
PLEX_SERVER = 'http://127.0.0.1:32400'  # Address of Plex server to query
USERNAME = ''  # Plex username to login, only required on first run
PASSWORD = ''  # Plex password to login, after first run, stored in keyring

# The location of the file that stores Plex token
token_path = os.path.expanduser(WORKING_DIR) + '.plextoken'

# Check if there's a token saved
token = None
username = None
if os.path.exists(token_path):
    with open(token_path) as token_file:
        token = token_file.readline().split(',')
    username, token = token
    token_last_update = os.path.getmtime(token_path)
    token_last_update = datetime.fromtimestamp(token_last_update)
    if (datetime.now() - token_last_update).days >= 1:
        token = None

# Update the token if there wasn't one saved, or it's older than a day
if not token:
    username = username or USERNAME
    password = keyring.get_password('Plex-Now-Playing', username) or PASSWORD
    keyring.set_password('Plex-Now-Playing', username, password)
    token_request = requests.post(
        url='https://plex.tv/users/sign_in.json',
        data='user%5Blogin%5D=' + username + '&user%5Bpassword%5D=' + password,
        headers={
            'X-Plex-Client-Identifier': 'Plex-Now-Playing',
            'X-Plex-Product': 'Plex-Now-Playing',
            'X-Plex-Version': '0.0.1'}
    )
    if token_request.status_code < 300:
        token = json.loads(token_request.text)['user']['authToken']
        token_data = {
            'username': username,
            'token': token,
            'last_update': datetime.now()
        }
        with open(token_path, 'w') as token_file:
            token_file.write('{0},{1}'.format(username, token))

    else:
        print('Authentication problem')

# Ask the local Plex server for the current sessions
if token:
    now_playing = requests.get(
        url=PLEX_SERVER + '/status/sessions',
        headers={'X-Plex-Token': token}
    )
    streams_xml = ET.fromstring(now_playing.text)  # Parse the XML returned by Plex

# For each stream, print an informative line about the stream
if len(streams_xml):  # Is len() here necessary?
    for stream in streams_xml:

        # Return the user watching the stream.
        user_full = str(stream.find('User').get('title', default='Unknown User'))
        user = re.search('[^@]+', user_full)
        if user is not None:
            user = user.group(0)
        else:
            user = user_full

        # Return either the time a transcode started _or_ the current time as a string
        start_time = stream.get('lastViewedAt', default=datetime.now())
        if type(start_time) is str:
            start_time = datetime.fromtimestamp(int(start_time))
        start_time = start_time.strftime('%b %d, %H:%M')  # Mon 01, HH:MM

        # The regex returns None if there's no match, None.group() is AttributeError
        season = re.search('[0-9]+', str(stream.get('parentTitle')))
        if season is not None:
            season = season.group(0)

        # Truncate the series name to make it shorter
        series_name_long = stream.get('grandparentTitle', default='Unknown Series')
        series_name = ' '.join(series_name_long.split(' ')[:5])  # Only first five words

        # Define how each type of stream is displayed
        stream_formats = dict()
        stream_formats['episode'] = '{start}: {user} // {series} - S{season}E{episode} - {tv_name}'
        stream_formats['movie'] = '{start}: {user} // {movie_title} ({year})'
        stream_formats['track'] = '{start}: {user} // {track_artist} - {track_title}'
        stream_formats.setdefault('{start}: {user} // Unknown Stream')

        # Apply formatting to stream data
        display_str = stream_formats[stream.get('type')].format(
            start=start_time,
            movie_title=stream.get('title', default='Unknown Movie'),
            user=user,
            series=series_name,
            season=season,
            episode=stream.get('index', default=''),
            tv_name=stream.get('title', default='Unknown Episode'),
            year=stream.get('originallyAvailableAt').split('-')[0],
            track_title=stream.get('title', default='Unknown Song'),
            track_artist=stream.get('grandparentTitle', default='Unknown Artist'),
            track_album=stream.get('parentTitle', default='Unknown Album'),
        )

        print(display_str[:75])  # Truncate display at 75 characters
else:
    print('Nothing playing')

File: plex-now-playing.R
# Set working directory (ending with /) where it can save tokens, find credentials.
# NOTE: Password will be saved in plaintext in this folder! Choose somewhere secure.
wkdir <- "~/Scripts/"

# Set the path to the credentials and the auth token
credential_path <- paste0(wkdir, ".plex_creds")
token_path <- paste0(wkdir, ".plex_token")

# To automatically write the .plex_creds file, uncomment and modify the following lines 
# if (!(file.exists(credential_path))) {
#   write.table(data.frame("username" = "John_k_User", 
#                          "password" = "hunter2"),
#               file = paste0(wkdir, ".plex_creds"), 
#               col.names = FALSE, row.names = FALSE, sep = ",")
# }
# The above lines can be commented out after the first run

# Load the necessary packages
packages <- c("dplyr", "tibble", "httr", "readr", "tidyr", "stringr")
if (!all(unlist(lapply(packages, require, character.only = TRUE)))) {
  install.packages(packages)
  lapply(packages, require, character.only = TRUE)
}

# Check if there's a token written and read it in
if (file.exists(token_path)) {
  token <- read_csv(token_path)
} else {
  token <- NULL
}

# If there's no token or the token is more than a day old, re-auth with plex.tv and get a new token
if (!length(token) | token$authed_at < Sys.Date()) {
  credentials <- read_csv(credential_path, col_names = FALSE)
  
  token_response <- content(
    POST(
      url = "https://plex.tv/users/sign_in.json", 
      body = paste0("user%5Blogin%5D=", credentials[[1]], "&user%5Bpassword%5D=", credentials[[2]]), 
      add_headers("X-Plex-Client-Identifier" = "NowPlayingScript", 
                  "X-Plex-Product" = "NowPlayingScript",
                  "X-Plex-Version" = "0.0.1")
    )
  )
  
  if (length(token_response$error)) {
    cat("Error getting authentication token! Check Plex credentials")
    quit(status = 1)
  }

  token <- token_response %>% 
    tibble(object = .) %>% 
    unnest_wider(object) %>%
    select(-subscription, -roles) %>% 
    add_column(authed_at = Sys.Date())
  
  write_csv(token, path = token_path) 
  
}

token <- token$authToken

# Get data on the current streams from the local Plex server
now_playing <- content(GET("http://127.0.0.1:32400/status/sessions", add_headers("X-Plex-Token" = token)))

# Check if something is playing, and if it is, print some data about the streams
if (!is.null(now_playing$MediaContainer$Metadata)) {
  np_nested <- tibble(metadata = now_playing$MediaContainer$Metadata) %>%
    unnest_wider(metadata) %>%
    unnest_wider(User, names_sep = ".")
  
  # If there's no "last viewed at" time, use the current time
  if (is.null(np_nested$lastViewedAt)) np_nested$lastViewedAt <- as.numeric(Sys.time())
  
  current <- np_nested %>%
    rowwise() %>%
    mutate(episode = paste0("S", gsub("[A-Za-z ]+", "", parentTitle), "E", str_pad(index, 2, pad = "0"))) %>%
    select(user = User.title,
           series = grandparentTitle,
           ep_name = title,
           time = lastViewedAt,
           episode)
  
  print_pretty <- function(df, row) {
    time <- as.numeric(df$time[row]) %>% 
      as.POSIXct(origin = "1970-01-01", tz = Sys.timezone()) %>% 
      format("%b %d, %I:%M%P")
    paste0(time, ": ", 
           substr(df$user[row], start = 1, stop = 12), " // ", 
           substr(df$series[row], start = 1, stop = 20), " - ", 
           substr(df$episode[row], start = 1, stop = 20), " - ", 
           substr(df$ep_name[row], start = 1, stop = 20)
    )
  }
  
  for (i in 1:nrow(current)) {
    cat(print_pretty(current, i))
    cat("\n")
  }
} else {
  cat("Nothing playing")
}
